### PR TITLE
Upgrade our build to sbt 1.5.4.

### PR DESCRIPTION
**Pull request by sjrd**
Naively upgrading to sbt >= 1.3.x caused lots of spurious recompilations. (See past attempt in #4360.) After some investigation, it was caused by our `ExternalCompile.scala` setup, which overrides `compile` but not `compileIncremental`. Overriding `compileIncremental` proved virtually impossible. Instead, I looked into getting rid of `ExperimentalCompile` altogether, which led me to several changes that were necessary to play nicely with the JVM back-end and the incremental compiler.

### CODE REVIEW

**Review by gzm0 — APPROVED** [review]

File: linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/VarGen.scala [hunk 1]
```diff
@@ -208,11 +208,23 @@ private[emitter] final class VarGen(jsGen: JSGen, nameGen: NameGen,
...
     typeRef match {
       case primRef: PrimRef =>
+        implicit val primRefScope: Scope[PrimRef] = Scope.PrimRefScope
         globalVar(field, primRef)
 
       case ClassRef(className) =>
+        implicit val classScope: Scope[ClassName] = Scope.ClassScope
         globalVar(field, className)
     }
   }
```
Comment (sjrd) [inline linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/VarGen.scala:230]: @gzm0 I had to add the above changes to make the CI pass. There were errors like
```
[error] linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/VarGen.scala:213:18: could not find implicit value for evidence parameter of type VarGen.this.Scope[org.scalajs.ir.Types.PrimRef]
[error] Error occurred in an application involving default arguments.
[error]         globalVar(field, primRef)
[error]                  ^
[error] linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/VarGen.scala:216:18: could not find implicit value for evidence parameter of type VarGen.this.Scope[org.scalajs.ir.Names.ClassName]
[error] Error occurred in an application involving default arguments.
[error]         globalVar(field, className)
[error]                  ^
[error] two errors found
[error] Scaladoc generation failed
```
When trying to investigate, I also observed the same error on my local Windows machine, but only in the IDE. That was without regenerating the build, so supposedly the IDE was still using the config generated through sbt 1.2.8.

Is that OK for you?

Comment (gzm0) [reply linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/VarGen.scala:230]: If it's the only thing we can do :-/ In the past for me, the incremental compiler failed on this specific thing if there were other errors in the same file. IIRC, cleaning helped.

Are we 100% sure the CI environment is clean?

Comment (sjrd) [reply linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/VarGen.scala:230]: We do `git clean -fdx && rm -rf partest/fetchedSources/` at the beginning of each CI job, so yes, it's pretty clear that it is clean.

Comment (sjrd) [reply linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/VarGen.scala:230]: Also `linker$v/compile` succeeds, during the same CI run, before `linker$v/compile:doc` fails, so ...

**Review by gzm0 — APPROVED** [review]

**Pull request merged**